### PR TITLE
Produce a .version file for Microsoft.NETCore.App in the runtime layout.

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Fixes an issue with VS not showing .NET 6 in the TFM dropdown for a new project.

### CODE REVIEW

File: src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Runtime.sfxproj [hunk 1]
```diff
@@ -121,6 +121,21 @@
     <NativeRuntimeAsset Include="$(_diaSymReaderPathIfExists)" />
   </ItemGroup>
 
+  <!-- VS uses this file to show the target framework in the drop down. -->
+  <Target Name="CreateDotVersionFile" DependsOnTargets="InitializeSourceControlInformationFromSourceControlManager" BeforeTargets="GetFilesToPublish">
```
Comment (ViktorHofer) [inline src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Runtime.sfxproj:125]: Please condition this target on `$(DisableSourceLink) != true` similar to what we do for the native version file: https://github.com/dotnet/arcade/blob/2b430e5bbfaec37a6cead2f0cf79157f01f1a623/src/Microsoft.DotNet.Arcade.Sdk/tools/Version.targets#L47.

File: src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Runtime.sfxproj [hunk 2]
```diff
@@ -121,6 +121,21 @@
...
+    <ItemGroup>
+      <_VersionFile Include="$(IntermediateOutputPath).version" TargetPath="shared/$(SharedFrameworkName)/$(Version)/" />
+    </ItemGroup>
+    <WriteLinesToFile
+      Lines="$(SourceRevisionId);$(Version)"
+      File="@(_VersionFile)"
+      Overwrite="true"
+      WriteOnlyWhenDifferent="true" />
+    <ItemGroup>
+      <FilesToPublish Include="@(_VersionFile)" />
```
Comment (ViktorHofer) [inline src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Runtime.sfxproj:135]: Also add the generated file to FileWrites?

File: src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Runtime.sfxproj (hunk 1 above)
Comment (hoyosjs) [inline src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Runtime.sfxproj:125]: Is this not needed all up? aka, will this fit better on the SFX sdk. Also, this feels like a really soft contract with VS. Why do they need said file? And can this be tested for, or added as a check in the SFX sdk?

File: src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Runtime.sfxproj (hunk 2 above)
Comment (hoyosjs) [inline src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Runtime.sfxproj:135]: Add to file writes too

File: src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Runtime.sfxproj (hunk 1 above)
Comment (jkoritzinsky) [reply src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Runtime.sfxproj:125]: @dsplaisted can speak more to the implicit contract with VS and if we need to do this all up in the SFX SDK.

Comment (dsplaisted) [reply src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Runtime.sfxproj:125]: I don't know the details, it has to do with how the template engine works.  @phenning or someone from @donJoseLuis's team can probably explain more about it.

It does seem like it should go in the shared framework SDK, as it looks like all 3 shared frameworks in .NET 5 had the .version file.  But if we want to fix this for .NET 6 preview 1 then we might want a quicker change in this repo for that.

File: src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Runtime.sfxproj (hunk 2 above)
Comment (jkoritzinsky) [reply src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Runtime.sfxproj:135]:
```suggestion
      <FilesToPublish Include="@(_VersionFile)" />
      <FileWrites Include="@(_VersionFile)" />
```

File: src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Runtime.sfxproj (hunk 1 above)
Comment (jkoritzinsky) [reply src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Runtime.sfxproj:125]:
```suggestion
  <Target Name="CreateDotVersionFile"
          DependsOnTargets="InitializeSourceControlInformationFromSourceControlManager"
          BeforeTargets="GetFilesToPublish"
          Condition="'$(DisableSourceLink)' != 'true'">
```

**Review by hoyosjs — APPROVED** [review]
It this is needed for 6.0p1, this looks good. Just remains the question of the contract with VS and if it should be in the SFX SDK, but that shouldn't block this.

**Review by ViktorHofer — APPROVED** [review]
Looks good, thanks. Let's make sure that we follow-up on moving this into the sfx sdk if appropriate.

**Review by safern — APPROVED** [review]

**Comment by phenning** [comment]
Some historical context:

The reason we check for this file is to make sure that we have a runtime installed for corresponding templates so that users don’t run into a situation where templates targeting a particular SDK are installed with the SDK or Visual Studio but no runtime was present which would cause F5 to fail.

We don’t simply check for the folder, but instead this file since it we’ve seen the folder left behind in the past due to setup uninstall issues.


Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by hoyosjs** [comment]
> Some historical context: The reason we check for this file is to make sure that we have a runtime installed for corresponding templates so that users don’t run into a situation where templates targeting a particular SDK are installed with the SDK or Visual Studio but no runtime was present which would cause F5 to fail. We don’t simply check for the folder, but instead this file since it we’ve seen the folder left behind in the past due to setup uninstall issues.

Good to know. If this was just for runtime, then I'd recommend using a file that needs to be present, like `System.Private.Corelib.dll`, but if this is a generic check for F5 on all project types, then SFX SDK should have this as a contract for interaction with VS.

**Comment by jkoritzinsky** [comment]
I'll start the process of backporting this to 6.0p1 tomorrow morning first thing. And start working on integrating this into the SDK in Arcade for master after that.

**Pull request merged**